packages/ecc-client-ga4gh-tes/src/components/tes-create-run/definition/tesCreateRun.ts

```diff
@@ -0,0 +1,544 @@
+  `,
+  {positioning : true}
+  )}
+  ````
+  ``` Ts
+  handleDataFieldChange = (event: Event, index: number) => {
+    this.data[index] = (event.target as HTMLInputElement).value;
+  };
+  ```
+  This would be the ideal way to handle the above mentioned rules, note this way uses
+  `this` in the class method and no params are being reassigned.
+
+But as mentioned in the above, the code uses { positioning: true } to get the index value, this as mentioned
+in the [FAST documentation](https://www.fast.design/docs/fast-element/using-directives/#the-repeat-directive:~:text=Some%20context%20properties,template%20from%20above%3A)
+can have performance issues, hence these rules are opted out.
+*/
+import {
+  FASTElement,
+  attr,
+  observable,
+  customElement,
+} from '@microsoft/fast-element';
+import template from './tesCreateRun.template.js';
+import styles from './tesCreateRun.styles.js';
+import { ExecutorData, InputData, OutputData } from './createTask.js';
+import { postTask } from '../../../data/Task/tesGet.js';
+
+const executorTemplate: ExecutorData = {
+  command: [''],
+  env: {},
+  image: '',
+  stderr: '',
+  stdin: '',
+  stdout: '',
+  workdir: '',
+};
+
+const inputTemplate: InputData = {
+  path: '',
+  url: '',
+};
+
+const outputTemplate: OutputData = {
+  path: '',
+  url: '',
+  type: '',
+};
+@customElement({
+  name: 'ecc-client-ga4gh-tes-create-run',
+  template,
+  styles,
+  shadowOptions: { mode: 'open' },
+})
+export default class TESCreateRun extends FASTElement {
+  @attr baseURL = '';
+
+  @attr name = '';
+
+  @attr description = '';
+
+  @attr taskExecutors: ExecutorData[] = [{ ...executorTemplate }];
+
+  @observable taskExecutorsLength = this.taskExecutors.length;
+
+  @attr taskInput: InputData[] = [{ ...inputTemplate }];
+
+  @observable taskInputLength = this.taskInput.length;
+
+  @attr taskOutput: OutputData[] = [{ ...outputTemplate }];
+
+  @observable taskOutputLength = 1;
+
+  @attr cpu_cores = '';
+
+  @attr disk_gb = '';
+
+  @attr preemptible = false;
+
+  @attr ram_gb = '';
+
+  @attr zones: string[] = [];
+
+  @attr WORKFLOW_ID = '';
+
+  @attr PROJECT_GROUP = '';
+
+  @attr volumes: string[] = [];
+
+  @observable response = {};
+
+  @observable taskData: any = {
+    name: this.name,
+    description: this.description,
+    executors: this.taskExecutors,
+    inputs: this.taskInput,
+    outputs: this.taskOutput,
+    resources: {
+      cpu_cores: this.cpu_cores,
+      disk_gb: this.disk_gb,
+      preemptible: this.preemptible,
+      ram_gb: this.ram_gb,
+      zones: this.zones,
+    },
+    tags: {
+      WORKFLOW_ID: this.WORKFLOW_ID,
+      PROJECT_GROUP: this.PROJECT_GROUP,
+    },
+    volumes: this.volumes,
+  };
+
+  connectedCallback(): void {
+    super.connectedCallback();
+    this.taskExecutorsLength = this.taskExecutors.length;
+    this.taskInputLength = this.taskInput.length;
+    this.taskOutputLength = this.taskOutput.length;
+  }
+
+  removeEmptyFields = (obj: any) => {
```

Comment:
this is one hell of a function, but it seems like  a bit of an over complication imo. Can you walk me through why you chose this approach instead of just creating another object and copying over the fields that have been filled?

Reply:
yeah sure,

See how the API is working is that there is no set data type that it accepts, ie, apart from
`{
  "executors": [
    {
      "command": [
        ""
      ],
      "image": "ubuntu:20.04",
      "workdir": "/data/"
    }
  ]
}`

All other fields are optional, now which means that, except these all other fields are not required right, well wrong, is let's say x nonessential field is left empty then the API will throw an error; OK, now let's say the user wants to fill 3 fields and leave rest empty then it is clients job to remove them from the API request body otherwise the API will throw an error.

I think it will be quite difficult to create data types for each possibility, i.e., one where 1 nonessential field is left out, then one where two are left out, and so forth.

This is the problem I have been trying to tell you. I hope this makes it clear. I understand this is quite a complex way to do it but I I could find another way, please feel free to correct me or re-write this.

@git-anurag-hub can you also look into this, please? I'd like to know your thoughts on this approach to this problem.

Reply:
@SalihuDickson Please look into this [API](https://csc-tesk-noauth.rahtiapp.fi/swagger-ui.html#/TaskService/createTask); maybe there is something I do not yet understand/ am overlooking in the way that the API is working, try creating a task with empty stderr and then try creating the task with no stderr in the object, and then do the same thing with various other keys. :+1:

Reply:
I understand what you're saying, @JaeAeich. However, I believe that @SalihuDickson's suggestion was to create a new object and add the key-value pairs to it instead of removing them from the previous object. Something like this maybe
```
const removeEmptyFields = (obj) => {
  const newObj = Array.isArray(obj) ? [] : {};

  Object.entries(obj).forEach(([key, value]) => {
    if (Array.isArray(value)) {
      const cleanedArray = value.map((element) => {
        if (typeof element === 'object' && element !== null) {
          return removeEmptyFields(element);
        }
        return element;
      }).filter((element) => {
        return !(typeof element === 'object' && Object.entries(element).length === 0);
      });

      if (cleanedArray.length > 0) {
        newObj[key] = cleanedArray;
      }
    } else if (value && typeof value === 'object') {
      const cleanedValue = removeEmptyFields(value);
      if (Object.entries(cleanedValue).length > 0) {
        newObj[key] = cleanedValue;
      }
    } else if (value !== null && value !== undefined && value !== '') {
      newObj[key] = value;
    }
  });

  return newObj;
};
```

Could you please confirm @SalihuDickson? I'm not entirely sure if there are any significant advantages to it, other than immutability and safety. However, the previous implementation is more space-efficient and performs better. I'm comfortable with either implementation.

Reply:
If this is what was asked, and both of you agree, I think I should do it; my concern, apart from space and efficiency would be state management; as the number of states grows, it will be hard for the package to be maintained later on.

Reply:
As said I am okay with both options. But let's wait for @SalihuDickson opinions on this. I guess, then we can merge this without an issue.

Reply:
okay, i'm not going to hold this up any longer. To be clear though i never thought the implementation was faulty, i just always favor simplicity so i like to understand when a more complex solution might be necessary.